### ISSUE TITLE
ES: Allow deconstructing objects with splats

### ISSUE THREAD

**Issue by flying-sheep** [jashkenas/coffeescript#3894]
…for parity with ECMA7. currently we can’t do this:

``` coffeescript
x =
  a: 1
  b: 2
  c: 3

{ a, b, y... } = x
# or
{ a, b, ...y } = x

console.log a, b, y  # → 1 2 {c:3}
```

**Comment by vendethiel** [comment]
Seems like #815 proposed this feature but was closed for other reasons.

**Comment by lydell** [comment]
Is that really part of ES6? I've never seen it before.

**Comment by flying-sheep** [comment]
ah, you seem to be right, it’s actually part of [JSX](http://facebook.github.io/react/docs/transferring-props.html#transferring-with-...-in-jsx). i assumed everything of JSX that isn’t `<ReactElement/>` syntax is a subset of ES6.

dayum, i think it’s really useful

**Comment by lydell** [comment]
Cool! Thanks for those links!

**Comment by flying-sheep** [comment]
no problem.

at closing of that other issue: luckily, that’s about something different (i.e. extending objects, not destructuring them)

and @jashkenas’ argument…

> Splats are for lists of things

…is also not valid anymore now that ECMAscript itself uses splats for objects.

so this should be good to go!

**Comment by flying-sheep** [comment]
now it’s in the [compat-table](http://kangax.github.io/compat-table/es7/#Object_rest_properties)

**Comment by bbtfr** [comment]
+1 for this, it would be very helpful to write React Apps in CoffeeScript with this feature.

**Comment by reduxdj** [comment]
Hoping this is added to Coffeescript, is it going to be?

**Comment by wmertens** [comment]
So I'd also love to see `{b:5, bar..., c}` as the shorthand for `merge({b: 5}, bar, {c: c})` and while we're at it `{[foo]: bar}` for `_t={},_t[foo]=bar,_t`, all very handy things slated for ES6/7.

Should these go into separate issues?

**Comment by lydell** [comment]
`{[foo]: bar}` is already possible: `{"#{foo}": bar}`

**Comment by wmertens** [comment]
Ooh, I didn't know about that, nicer syntax too! That just leaves the spread and rest operations...

**Comment by michaelficarra** [comment]
@lydell with the caveat that it doesn't work for symbols.

**Comment by vendethiel** [comment]
Well, since @jashkenas said he considered symbols a misfeature, I guess that's fine for coffee?

**Comment by michaelficarra** [comment]
Guess so, though I don't agree we can ignore it since standard lib APIs and even syntax depend on values with symbols as keys.

**Comment by flying-sheep** [comment]
yeah, we need a way to express this ES2015 in CS:

```js
const iterable = {
	*[Symbol.iterator]() {
		yield 1
		yield 2
	}
}
```

**Comment by wmertens** [comment]
I'd just like to point out that I'm currently mostly writing JSX and that's mostly because of being able to write `foo={...props, thing: 5}`.

**Comment by onemanstartup** [comment]
@wmertens https://github.com/jsdf/coffee-react cofee-react supports spread attributes.

**Comment by wmertens** [comment]
@onemanstartup yeah but I don't want to write xml.
I simply set `E` to `React.createElement` and then write things like `E Menubar, collapse: false, (E Item label: t.name, t.val for t in items)`.
Just calling functions is very nice in CS, not so much is JS.

**Comment by Jamesernator** [comment]
> yeah, we need a way to express this ES2015 in CS:
> 
> const iterable = {
>     *[Symbol.iterator]() {
>         yield 1
>         yield 2
>     }
> }

This is something that should probably be an issue of its own as currently the only ways to do this are:

Not inheritable
```coffeescript
class SomeIterable
    constructor: (...) ->
        ...
        @[Symbol.iterator] ->
            ...
```

Sort've defeats the original point of CoffeeScript classes:
```coffeescript
class SomeIterable
    constructor: (...) ->

SomeIterable::[Symbol.iterator] = ->
    ...
```

**Comment by lydell** [comment]
@Jamesernator This is a bit nicer:

```coffee
class SomeIterable
    constructor: (...) ->

    @::[Symbol.iterator] = ->
        ...
```

**Comment by VoronaPavel** [comment]
Speaking of cjsx, one of the main feature of object splats (or rest parameters) in JSX is ability to do [so](https://facebook.github.io/react/docs/transferring-props.html#transferring-with-...-in-jsx):
```js
const {propIamGonaConsume, ...others} = props
return <SomeComponent {...others}/>
```
As it seems to me, for now it is impossible to do in cjsx, so it would be really nice to resolve this feature request.

**Comment by VoronaPavel** [comment]
Also, it's first time i found that JS can do something coffee can't

**Comment by Tsher** [comment]
Just use ES6 emit instead. may be coffeescript is good support to ES6 ,not ES7

	obj = { x: 1, y: 2}
	obj1 = {...obj, z: 3}

	# ES6 emit
	obj = {x: 1, y: 2}
	obj1= Object.assign({}, obj, { z: 3 })

**Comment by dbackeus** [comment]
As immutable object creation in redux etc is becoming all the rage replicating the object splat in coffeescript seems quite relevant.

It would certainly be a nicer solution than using `Object.assign`, which is also polyfill dependent.

**Comment by gucki** [comment]
@jashkenas Would a PR get merged?

**Comment by vendethiel** [comment]
It'd certainly start a discussion :-).

**Comment by beeplin** [comment]
+1.
```
obj = { x: 1, y: 2}
obj1 = {...obj, z: 3}
```
nowadays allowing this grammar is quite vital when writing react/redux and vue/vuex. hoping coffee is catching this soon. :)

**Comment by benjie** [comment]
I'd be fine with this being simply passed through so Babel can interpret it, like `async`/`await`

**Comment by GeoffreyBooth** [comment]
This should be supported as part of CoffeeScript 2, which will output ES2015 spread syntax. After #4311 implements the rest operator, spread in objects should probably be the next target. Help wanted.

**Comment by beeplin** [comment]
@GeoffreyBooth,  glad to see plans on CoffeeScript 2 ~ Some es6 features do become more and more vital today: `import` and `export` will be the key point of tree shaking in rollup and webpack2; some frameworks like react and vue are also designing their syntax to make max use of es6 feathers like object spread. We all like coffee and hope the new version of coffee can be going ahead of es6 again. :)

**Comment by GeoffreyBooth** [comment]
`import` and `export` are already here, in CS 1.11.

**Comment by beeplin** [comment]
wow, didn't notice that. cool!

**Comment by GeoffreyBooth** [comment]
Help wanted! I’ve started a [destructuring branch](https://github.com/GeoffreyBooth/coffeescript/tree/destructuring) for CoffeeScript 2 to output ES2015+ syntax for destructuring. I’m happy to invite contributors to this branch.

**Comment by GeoffreyBooth** [comment]
Implemented via #4493.
